Issue-1488: Quick start guide to developing in OpenMetadata (#1532)
open-metadata/OpenMetadata: docs/open-source-community/developer/quick-start-guide.md

# Quick start guide

Use this document as a quick start guide to begin developing in OpenMetdata. The document answers `what's` and `where's` of the following topic,

1. Schema, aka models 
2. API's 
3. System and Components

> ## Schema (aka, models)

Author: agreed.

OpenMetadata takes the schema-first approach to model metadata. The schema defines the entities, types, API requests, and relationships between the entities. The schema is written in JSON format following https://json-schema.org/ guidelines.  

The JSON Schemas are converted into [Java POJO's](https://www.jsonschema2pojo.org/) using `jsonschema2pojo-maven-plugin` plugin as defined in [`pom.xml`](https://github.com/open-metadata/OpenMetadata/blob/16d8ba548d968c09e6634eefbd32c87c66996b90/catalog-rest-service/pom.xml#L395). The generated `POJO's` are located under `OpenMetadata/catalog-rest-service/target/generated-sources/jsonschema2pojo`.

### Entities  
The entities are located at [`OpenMetadata/catalog-rest-service/src/main/resources/json/schema/entity`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/resources/json/schema/entity), and the following are few examples of the entities supported by OpenMetadata.
- data
- feed
- policies
- services
- tags
- teams

> ### Common Types    

Author: sure, changed to "Types".

All OpenMetadata supported types are defined under [`OpenMetadata/catalog-rest-service/src/main/resources/json/schema/type`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/resources/json/schema/type)

### API request objects  
The API request objects are defined under [`OpenMetadata/catalog-rest-service/src/main/resources/json/schema/api`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/resources/json/schema/api).  

## API
[Dropwizard](https://www.dropwizard.io/) java framework is used to build the RESTful API's.  All the API's are defined under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/resources`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/resources). 
[Swagger](https://swagger.io/) is used for API documentation following the OpenAPI standards.

## System and Components
![system](https://user-images.githubusercontent.com/3944743/144579545-e51aa37e-93ed-4080-9f32-ce617029571f.png)

### Events  

> Changes to the entities are captured as `events` and are stored in the database and elastic search.

Reviewer: Are change events searchable?

> Changes to the entities are captured as `events` and are stored in the database and elastic search.

Author: yes. took your suggestion.

The event handlers are defined under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events) and are applied globally to any outgoing response using the `ContainerResponseFilter`

### Database  

> MySql is used to persist the entities and the repository code to interact with MySql is located under [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3).

Reviewer: ```suggestion
OpenMetadata uses MySQL for the metadata catalog. The catalog code is located in the directory [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3`](https://github.com/open-metadata/OpenMetadata/tree/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3).
```

The database entity tables are created using the command [`OpenMetadata/bootstrap/bootstrap_storage.sh`](https://github.com/open-metadata/OpenMetadata/blob/main/bootstrap/bootstrap_storage.sh). [Flyway](https://flywaydb.org/) is used for managing the database table versions.

> ### Elastic Search  

Reviewer: ```suggestion
### Elasticsearch  
```

> Entity change events are stored in elastic search. The [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/events/ElasticSearchEventHandler.java) is responsible for capturing the change events and updating es.

Reviewer: It seems misleading to describe the role of Elasticsearch as solely indexing change events.

Author: changed the doc, please let me know if this is ok.

OpenMetadata uses Elasticsearch to store the Entity change events and makes it searchable by search index. The [`OpenMetadata/catalog-rest-service/src/main/java/org/openmetadata/catalog/elasticsearch/ElasticSearchEventHandler.java`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/elasticsearch/ElasticSearchEventHandler.java) is responsible for capturing the change events and updating Elasticsearch.

Elasticsearch indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.

> The es indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.

Reviewer: ```suggestion
Elasticsearch indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.
```

### Authentication/Authorization  

> Authentication is provided by google oauth provider. All incoming requests are filtered by validating the JWT token using the google oauth provider. Access control is provided by [`CatalogAuthorizer`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/security/CatalogAuthorizer.java).

Reviewer: ```suggestion
OpenMetadata uses Google OAuth for authentication. All incoming requests are filtered by validating the JWT token using the Google OAuth provider. Access control is provided by [`CatalogAuthorizer`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/security/CatalogAuthorizer.java).
```

> Auth/Authz details are configured at [`OpenMetadata/conf/openmetadata-security.yaml`](https://github.com/open-metadata/OpenMetadata/blob/main/conf/openmetadata-security.yaml)

Reviewer: ```suggestion
See the configuration file [`OpenMetadata/conf/openmetadata-security.yaml`](https://github.com/open-metadata/OpenMetadata/blob/main/conf/openmetadata-security.yaml) for the authentication and authorization configurations.
```

### Ingestion  

> Ingestion is a simple python framework to ingest the metadata from various external sources into OpenMetadata platform.

Reviewer: ```suggestion
Ingestion is a simple Python framework to ingest metadata from external sources into OpenMetadata.
```

**Connectors**

> The ingestion is carried out using a set of components called `Connectors` as described [here](https://docs.open-metadata.org/open-source-community/developer/build-a-connector#workflow) and it consists of

Reviewer: ```suggestion
OpenMetadata defines and uses a set of components called `Connectors` for metadata ingestion. Each data service requires its own connector.  See the documentation on how to [build a connector](https://docs.open-metadata.org/open-source-community/developer/build-a-connector#workflow) for details on developing connectors for new services. 
```

There are some popular connectors already developed and can be found under

1. Source → [`OpenMetadata/ingestion/src/metadata/ingestion/source`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/source)
2. Processor → [`OpenMetadata/ingestion/src/metadata/ingestion/processor`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/processor)
3. Sink → [`OpenMetadata/ingestion/src/metadata/ingestion/sink`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/sink)
4. Stage → [`OpenMetadata/ingestion/src/metadata/ingestion/stage`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/stage)
5. BulkSink → [`OpenMetadata/ingestion/src/metadata/ingestion/bulksink`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/ingestion/bulksink)

**Airflow**

> For simplicity, the metadata from external sources are ingested in a pull-based model, and Apache Airflow is used as an orchestration framework to perform ingestion.

Reviewer: ```suggestion
For simplicity, OpenMetadata ingests metadata from external sources using a pull-based model. OpenMetadata uses Apache Airflow to orchestrate ingestion workflows.
```

> Example Airflow dags are found under [`OpenMetadata/ingestion/examples/airflow/dags`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/examples/airflow/dags) for reference.

Reviewer: ```suggestion
See the directory [`OpenMetadata/ingestion/examples/airflow/dags`](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/examples/airflow/dags) for reference DAG definitions.
```

**JsonSchema python typings**

> The python typings for OpenMetadata JsonSchema is generated from the [`Makefile`](https://github.com/open-metadata/OpenMetadata/blob/main/Makefile) using the command `make generate`. The generated files are located at `OpenMetadata/ingestion/src/metadata/generated`

Reviewer: ```suggestion
You can generate Python types for OpenMetadata models defined using Json Schema using the `make generate` command of the [`Makefile`](https://github.com/open-metadata/OpenMetadata/blob/main/Makefile). Generated files are located in the directory `OpenMetadata/ingestion/src/metadata/generated`
```

The es indices are created when the [`OpenMetadata/ingestion/pipelines/metadata_to_es.json`](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/pipelines/metadata_to_es.json) ingestion connector is run.

### Authentication/Authorization  
Authentication is provided by google oauth provider. All incoming requests are filtered by validating the JWT token using the google oauth provider. Access control is provided by [`CatalogAuthorizer`](https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/java/org/openmetadata/catalog/security/CatalogAuthorizer.java).

Auth/Authz details are configured at [`OpenMetadata/conf/openmetadata-security.yaml`](https://github.com/open-metadata/OpenMetadata/blob/main/conf/openmetadata-security.yaml)

### Ingestion  
Ingestion is a simple python framework to ingest the metadata from various external sources into OpenMetadata platform.

> Ingestion is a simple python framework to ingest the metadata from various external sources into OpenMetadata platform.

Reviewer 2: Duplicate line as the one above?

Author: thank you. fixed.
